src/r_exasol/ssl/certificate.cpp

```diff
@@ -0,0 +1,169 @@
+#include "certificate.h"
+#include <openssl/conf.h>
+#include <openssl/x509v3.h>
+#include <openssl/ssl.h>
+#include <string>
+
+//Code based on OpenSSL demo:
+//https://opensource.apple.com/source/OpenSSL/OpenSSL-22/openssl/demos/x509/mkcert.c
+
+
+namespace exa {
+    namespace ssl {
+
+        struct CertExtension {
+            CertExtension(int nid, const char* value) :
+            mNid(nid),
+            mValue(value) {}
+
+            int mNid;
+            const char *mValue;
+        };
+
+        class CertExtensionBuilder {
+        public:
+            explicit CertExtensionBuilder(X509 *cert) : mCert(cert) {}
+
+            CertExtensionBuilder& operator << (const CertExtension && certExtension);
+        private:
+            X509 *mCert;
+        };
+
+        /* Add extension using V3 code: we can set the config file as NULL
+         * because we wont reference any other sections.
+         */
+
+        CertExtensionBuilder &CertExtensionBuilder::operator<<(const CertExtension &&certExtension) {
+            X509_EXTENSION *ex;
+            X509V3_CTX ctx;
+            /* This sets the 'context' of the extensions. */
+            /* No configuration database */
+            X509V3_set_ctx_nodb(&ctx);
+            /* Issuer and subject certs: both the target since it is self signed,
+             * no request and no CRL
+             */
+            X509V3_set_ctx(&ctx, mCert, mCert, nullptr, nullptr, 0);
+            ex = X509V3_EXT_conf_nid(nullptr, &ctx, certExtension.mNid, certExtension.mValue);
+            if (!ex) {
+                throw CertificateException("Cannot add extension to certificate.");
+            }
+            X509_add_ext(mCert,ex,-1);
+            X509_EXTENSION_free(ex);
+            return *this;
+        }
+
+        struct CertName {
+            CertName(const char * field, int type, const char * bytes)
+            : mField(field)
+            , mType(type)
+            , mBytes(bytes) {}
+            const char * mField;
+            int mType;
+            const char * mBytes;
+        };
+
+        class CertNameBuilder  {
+        public:
+            CertNameBuilder(X509_NAME *name) : mName(name) {}
+
+            CertNameBuilder & operator <<(const CertName && certName);
+
+        private:
+            X509_NAME *mName;
+        };
+
+        CertNameBuilder &CertNameBuilder::operator<<(const CertName &&certName) {
+            const unsigned char* bytes = reinterpret_cast<const unsigned char *>(certName.mBytes);
+            int ret = X509_NAME_add_entry_by_txt(mName, certName.mField,
+                                                 certName.mType, bytes, -1, -1, 0);
+            if (1 != ret) {
+                throw CertificateException(std::string("cannot set name of certificate for field:") + certName.mField);
+            }
+            return *this;
+        }
+    }
+}
+
+void exa::ssl::Certificate::mkcert(int bits, int serial, int days) {
```

Comment:
What do we do with long-running connection or R sessions, can their certificate expire. In that case,  a singleton might be a problem.

Reply:
The expiration time is 365 days. I don't think in practice this will be a problem.